src/main/java/appeng/tile/misc/TileInscriber.java

```diff
@@ -580,4 +625,93 @@ private void setProcessingTime( final int processingTime )
 	{
 		this.processingTime = processingTime;
 	}
+
+	/**
+	 * This is an item handler that exposes the inscribers inventory while providing simulation capabilities that do not
+	 * reset the progress if there's already an item in a slot. Previously, the progress of the inscriber was reset when
+	 * another mod attempetd insertion of items when there were already items in the slot.
+	 */
+	private class ItemHandler implements IItemHandler
+	{
+
+		private final int insertSlot;
+
+		private final int extractSlot;
+
+		private ItemHandler( int insertSlot, int extractSlot )
+		{
+			this.insertSlot = insertSlot;
+			this.extractSlot = extractSlot;
+		}
+
+		@Override
+		public int getSlots()
+		{
+			return insertSlot != extractSlot ? 2 : 1;
+		}
+
+		@Override
+		public ItemStack getStackInSlot( int slot )
+		{
+			if( slot == 0 )
+			{
+				return inv.getStackInSlot( insertSlot );
+			}
+			else if( insertSlot != extractSlot && slot == 1 )
+			{
+				return inv.getStackInSlot( extractSlot );
+			}
+			return null;
+		}
+
+		@Override
+		public ItemStack insertItem( int slot, ItemStack stack, boolean simulate )
+		{
+			if( slot != 0 || stack == null )
+			{
+				return stack;
+			}
+
+			// If there's already an item stack in the slot, we don't allow insertion and don't do any other checks
+			if( inv.getStackInSlot( insertSlot ) != null )
+			{
+				return stack;
+			}
+
+			AdaptorIInventory adapter = new AdaptorIInventory( new WrapperInventoryRange( TileInscriber.this, insertSlot, 1, true ) );
+
+			if( simulate )
+			{
+				return adapter.simulateAdd( stack );
+			}
+			else
+			{
+				return adapter.addItems( stack );
+			}
+		}
+
+		@Override
+		public ItemStack extractItem( int slot, int amount, boolean simulate )
+		{
+			final int validExtractSlot = ( insertSlot == extractSlot ) ? 0 : 1;
+
+			if( slot != validExtractSlot || amount == 0 )
+			{
+				return null;
+			}
+
+			AdaptorIInventory adapter = new AdaptorIInventory( new WrapperInventoryRange( TileInscriber.this, extractSlot, 1, true ) );
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'adapter' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.orod.org/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)